File: sdks/kotlin/src/main/kotlin/io/erebor/sdk/Erebor.kt
package io.erebor.sdk

import android.content.Context
import io.erebor.sdk.auth.AuthManager
import io.erebor.sdk.models.EreborConfig
import io.erebor.sdk.models.EreborUser
import io.erebor.sdk.network.ApiClient
import io.erebor.sdk.storage.SecureTokenStore
import io.erebor.sdk.wallet.BiometricGate
import io.erebor.sdk.wallet.WalletManager
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map

/**
 * Main entry point for the Erebor SDK.
 * 
 * Usage:
 * ```kotlin
 * val config = EreborConfig(
 *     apiUrl = "https://api.erebor.io",
 *     appId = "your-app-id",
 *     loginMethods = listOf(LoginMethod.EMAIL, LoginMethod.GOOGLE)
 * )
 * 
 * Erebor.configure(context, config)
 * 
 * if (!Erebor.isAuthenticated) {
 *     // Show login UI
 * }
 * ```
 */
object Erebor {
    private var _context: Context? = null
    private var _config: EreborConfig? = null
    private var _apiClient: ApiClient? = null
    private var _tokenStore: SecureTokenStore? = null
    private var _authManager: AuthManager? = null
    private var _walletManager: WalletManager? = null
    
    /**
     * Configure the Erebor SDK with context and configuration.
     * This must be called before using any other SDK functionality.
     */
    fun configure(context: Context, config: EreborConfig) {
        _context = context.applicationContext
        _config = config
        
        _tokenStore = SecureTokenStore(context)
        _apiClient = ApiClient(config.apiUrl, _tokenStore!!, config.tokenPrefix ?: "erebor")
        _authManager = AuthManager(_apiClient!!, _tokenStore!!)
        
        // WalletManager will be created lazily when needed
    }
    
    /**
     * Check if the SDK has been configured.
     */
    val isConfigured: Boolean
        get() = _config != null && _apiClient != null
    
    /**
     * Check if a user is currently authenticated.
     */
    val isAuthenticated: Boolean
        get() {
            ensureConfigured()
            return _tokenStore?.loadTokens() != null
        }
    
    /**
     * Get the current authenticated user, or null if not authenticated.
     */
    val user: StateFlow<EreborUser?>
        get() {
            ensureConfigured()
            return _authManager!!.authState.map { it.user }
        }
    
    /**
     * Access authentication functionality.
     */
    val auth: AuthManager
        get() {
            ensureConfigured()
            return _authManager!!
        }
    
    /**
     * Access wallet management functionality.
     */
    val wallets: WalletManager
        get() {
            ensureConfigured()
            if (_walletManager == null) {
                val context = _context!!
                val biometricGate = BiometricGate(context)
                _walletManager = WalletManager(_apiClient!!, biometricGate)
            }
            return _walletManager!!
        }
    
    /**
     * Get the current configuration.
     */
    val config: EreborConfig
        get() {
            ensureConfigured()
            return _config!!
        }
    
    private fun ensureConfigured() {
        if (!isConfigured) {
            throw IllegalStateException("Erebor SDK not configured. Call Erebor.configure() first.")
        }
    }
}

File: sdks/swift/Sources/Erebor.swift
import Foundation
import UIKit

/// Main entry point for the Erebor SDK
/// Provides access to authentication, wallet management, and other SDK features
public class Erebor: ObservableObject {
    public static let shared = Erebor()
    
    // MARK: - Public Properties
    
    /// Current configuration
    public private(set) var config: EreborConfig?
    
    /// Authentication state
    @Published public var isAuthenticated: Bool = false
    
    /// Current authenticated user
    @Published public var user: EreborUser?
    
    /// Loading state
    @Published public var isLoading: Bool = false
    
    // MARK: - Internal Components
    
    internal var apiClient: APIClient?
    internal var authManager: AuthManager?
    internal var walletManager: WalletManager?
    internal var keychainStore: KeychainStore?
    internal var deviceKeyManager: DeviceKeyManager?
    
    private init() {
        // Private initializer to enforce singleton pattern
    }
    
    // MARK: - Configuration
    
    /// Configure the Erebor SDK with the provided settings
    /// - Parameters:
    ///   - apiUrl: Base URL of the Erebor API
    ///   - appId: Your application identifier
    ///   - config: Additional configuration options
    public func configure(apiUrl: String, appId: String, config: EreborConfig) {
        self.config = config.with(apiUrl: apiUrl, appId: appId)
        
        // Initialize core components
        self.keychainStore = KeychainStore(serviceIdentifier: "erebor-\(appId)")
        self.apiClient = APIClient(config: self.config!)
        self.authManager = AuthManager(apiClient: self.apiClient!, keychainStore: self.keychainStore!)
        self.walletManager = WalletManager(apiClient: self.apiClient!, keychainStore: self.keychainStore!)
        self.deviceKeyManager = DeviceKeyManager(keychainStore: self.keychainStore!)
        
        // Check existing authentication state
        Task {
            await checkAuthenticationState()
        }
    }
    
    // MARK: - Authentication State
    
    @MainActor
    private func checkAuthenticationState() async {
        guard let authManager = authManager else { return }
        
        isLoading = true
        
        do {
            let user = try await authManager.getCurrentUser()
            self.user = user
            self.isAuthenticated = true
        } catch {
            // User not authenticated or token expired
            self.user = nil
            self.isAuthenticated = false
        }
        
        isLoading = false
    }
    
    // MARK: - Quick Access Properties
    
    /// Access to authentication methods
    public var auth: AuthManager? {
        return authManager
    }
    
    /// Access to wallet operations
    public var wallet: WalletManager? {
        return walletManager
    }
    
    // MARK: - Convenience Methods
    
    /// Quick login with email
    /// - Parameter email: User's email address
    /// - Returns: OTP session for verification
    public func loginWithEmail(_ email: String) async throws -> OTPSession {
        guard let authManager = authManager else {
            throw EreborError.sdkNotConfigured
        }
        return try await authManager.loginWithEmail(email)
    }
    
    /// Quick logout
    public func logout() async throws {
        guard let authManager = authManager else {
            throw EreborError.sdkNotConfigured
        }
        
        try await authManager.logout()
        
        await MainActor.run {
            self.user = nil
            self.isAuthenticated = false
        }
    }
}

// MARK: - Configuration Extension

private extension EreborConfig {
    func with(apiUrl: String, appId: String) -> EreborConfig {
        return EreborConfig(
            apiUrl: apiUrl,
            appId: appId,
            loginMethods: self.loginMethods,
            chains: self.chains,
            appearance: self.appearance,
            tokenPrefix: self.tokenPrefix
        )
    }
}

// MARK: - Error Types

public enum EreborError: LocalizedError {
    case sdkNotConfigured
    case authenticationRequired
    case biometricNotAvailable
    case userCancelled
    case invalidConfiguration
    case networkError(String)
    case cryptographicError(String)
    case keychainError(String)
    
    public var errorDescription: String? {
        switch self {
        case .sdkNotConfigured:
            return "Erebor SDK not configured. Call configure() first."
        case .authenticationRequired:
            return "User authentication is required for this operation."
        case .biometricNotAvailable:
            return "Biometric authentication is not available on this device."
        case .userCancelled:
            return "Operation was cancelled by the user."
        case .invalidConfiguration:
            return "Invalid configuration provided."
        case .networkError(let message):
            return "Network error: \(message)"
        case .cryptographicError(let message):
            return "Cryptographic error: \(message)"
        case .keychainError(let message):
            return "Keychain error: \(message)"
        }
    }
}